src/pczt.rs

```diff
@@ -0,0 +1,406 @@
+//! PCZT support for Orchard.
+
+use std::collections::BTreeMap;
+use std::fmt;
+
+use getset::{Getters, MutGetters};
+use pasta_curves::pallas;
+use zcash_note_encryption::OutgoingCipherKey;
+use zip32::ChildIndex;
+
+use crate::{
+    bundle::Flags,
+    keys::{FullViewingKey, SpendingKey},
+    note::{ExtractedNoteCommitment, Nullifier, RandomSeed, Rho, TransmittedNoteCiphertext},
+    primitives::redpallas::{self, Binding, SpendAuth},
+    tree::MerklePath,
+    value::{NoteValue, ValueCommitTrapdoor, ValueCommitment, ValueSum},
+    Address, Anchor, Proof,
+};
+
+mod parse;
+pub use parse::ParseError;
+
+mod io_finalizer;
+
+mod prover;
+pub use prover::ProverError;
+
+mod signer;
+pub use signer::SignerError;
+
+mod tx_extractor;
+pub use tx_extractor::{TxExtractorError, Unbound};
+
+/// PCZT fields that are specific to producing the transaction's Orchard bundle (if any).
+///
+/// This struct is for representing Orchard in a partially-created transaction. If you
+/// have a fully-created transaction, use [the regular `Bundle` struct].
+///
+/// [the regular `Bundle` struct]: crate::Bundle
+#[derive(Debug, Getters, MutGetters)]
+#[getset(get = "pub")]
+pub struct Bundle {
+    /// The Orchard actions in this bundle.
+    ///
+    /// Entries are added by the Constructor, and modified by an Updater, IO Finalizer,
+    /// Signer, Combiner, or Spend Finalizer.
+    #[getset(get_mut = "pub")]
+    pub(crate) actions: Vec<Action>,
+
+    /// The flags for the Orchard bundle.
+    ///
+    /// This is set by the Creator. The Constructor MUST only add spends and outputs that
+    /// are consistent with these flags (i.e. are dummies as appropriate).
+    pub(crate) flags: Flags,
+
+    /// The sum of the values of all `actions`.
+    ///
+    /// This is initialized by the Creator, and updated by the Constructor as spends or
+    /// outputs are added to the PCZT. It enables per-spend and per-output values to be
+    /// redacted from the PCZT after they are no longer necessary.
+    pub(crate) value_sum: ValueSum,
```

Comment:
The mut getter here is actually for the Fun Third Option: mutating existing actions so that signatures can be applied. In the PCZT format PR I implemented all the roles as consuming / transforming, but the `orchard` APIs are what get used under the hood to do the transformations, and for the scale of PCZTs a mapping-based API with the necessary granularity would be ridiculously verbose and onerous to work with.

If we don't want to expose the dual capabilities of "modify existing Actions" and "add / remove Actions" here, then I can replace this with a method that returns `&mut [Action]` instead (hiding the `Vec` nature and only exposing the modification capability). That would be sufficient for now as the "add Actions" capability is currently provided by `Builder` (I have no plans to enable piece-wise PCZT building in this first pass).

Reply:
Hiding the `Vec` nature sounds like the right approach to me. Right now it presents a footgun, which we can eliminate.